server/tests/test_migrations.py

```diff
@@ -0,0 +1,145 @@
+        print(f"run: {' '.join(cmd)}")
+        process = await asyncio.create_subprocess_exec(*cmd, stdout=asyncio.subprocess.PIPE)
+        stdout, stderr = await process.communicate()
+        return stdout.decode()
+
+    async def dump_data() -> str:
+        cmd = [pg_dump, "--schema=public", "--data-only", pg_cluster_url]
+        print(f"run: {' '.join(cmd)}")
+        process = await asyncio.create_subprocess_exec(*cmd, stdout=asyncio.subprocess.PIPE)
+        stdout, stderr = await process.communicate()
+        return stdout.decode()
+
+    async def restore_data(data: str) -> None:
+        cmd = [pg_psql, "--no-psqlrc", "--set", "ON_ERROR_STOP=on", pg_cluster_url]
+        print(f"run: {' '.join(cmd)}")
+        process = await asyncio.create_subprocess_exec(*cmd, stdin=asyncio.subprocess.PIPE)
+        await process.communicate(data.encode())
+
+    # The schema may start with an automatic comment, something like:
+    # `COMMENT ON SCHEMA public IS 'standard public schema';`
+    # So we clean everything first
+    await reset_db_schema()
+
+    # Now we apply migrations one after another and also insert the provided data
+    migrations = retrieve_migrations()
+    for migration in migrations:
+        result = await apply_migrations(pg_cluster_url, [migration], dry_run=False)
+        assert not result.error
+
+    # Save the final state of the database schema
+    schema_from_migrations = await dump_schema()
+
+    # Also save the final data
+    data_from_migrations = await dump_data()
+
+    # Now drop the database clean...
+    await reset_db_schema()
+
+    # ...and reinitialize it with the current datamodel script
+    await execute_datamodel()
+
+    # The resulting database schema should be equivalent to what we add after
+    # all the migrations
+    schema_from_init = await dump_schema()
+    assert schema_from_init == schema_from_migrations
+
+    # Final check is to re-import all the data, this requires some cooking first:
+
+    # Remove the migration related data given their should already be in the database
+    data_from_migrations = re.sub(
+        r"COPY public.migration[^\\]*\\.", "", data_from_migrations, flags=re.DOTALL
+    )
+    # Modify the foreign key constraint between `user_` and `device` to be
+    # checked at the end of the transaction. This is needed because `user_`
+    # table is entirely populated before switching to `device`. So the
+    # constraint should break as soon as an `user_` row references a device_id.
+    data_from_migrations = f"""
+ALTER TABLE public."user_" ALTER CONSTRAINT fk_user_device_user_certifier DEFERRABLE;
+ALTER TABLE public."user_" ALTER CONSTRAINT fk_user_device_revoked_user_certifier DEFERRABLE;
+
+BEGIN;
+
+SET CONSTRAINTS fk_user_device_user_certifier DEFERRED;
+SET CONSTRAINTS fk_user_device_revoked_user_certifier DEFERRED;
+
+{data_from_migrations}
+
+COMMIT;
+"""
+
+    await restore_data(data_from_migrations)
```

Comment:
I'll add a small comment: 
```suggestion
    # All the data should be restored without errors (restore_data will fail on first error)
    await restore_data(data_from_migrations)
```

Reply:
05bf51a06892c1295ab5a4648050882b124bed09